### ISSUE TITLE
Changing language of TGM plugin

### ISSUE THREAD

**Issue by geoclaps** [TGMPA/TGM-Plugin-Activation#506]
Can't seem to manage the language files to work within Wordpress (Dutch). I have copied the code above in a new pot file (and also a .po file) and exported it to an .mo file. Than uploaded it to /tgm/langauges/. Language keeps being set to English.

I also tried to setup custom language strings in my functions.php but after that there are still several words not being translated.

I use the plugin to include some plugins from the Wordpress.org respository so my theme always picks the most updated version of the plugins.

So, where to place the po or pot and mo files?

**Comment by jrfnl** [comment]
Hi @geoclaps, thank you for your willingness to create a translation for TGMPA. Much appreciated.

Based on your message, I get the impression that you are not very familiar with how localization works for WordPress. No offense meant.

So let me try to explain some of the basic concepts. Some of this you'll probably already know, but just to make sure you have all the information you need to get it working (the most important bit is in the last bullet in case you're thinking _"I know all this already"_):
* First off, a plugin or theme has to be _"translation ready"_ for you to be able to start translating. That means that all text strings use the WordPress _localization_ functions. For TGMPA, that is the case.
  
* A `.pot` file is like a 'master template' containing all language strings for a theme/plugin/core component. Generally speaking, you should not create/generate this yourself (unless it's for your own code), but rather take the official version from the library to start with. If you detect that that file is out of date, you could ask the theme/plugin author to update it.
   For TGMPA, you already noticed that I send in a PR to update the file, so you know were to find the up-to-date version ;-).
  
* Next, to start a translation, you create a new `.po` (=translation) file. POEdit is an excellent tool for this. On the start screen, just choose the second option _"Create new translation (from an existing .pot file)"_, Open the `tgmpa.pot` file you downloaded here, choose the language and start translating.
  While translating, keep an eye out for "context" information, plurals (easily overlooked) and sometimes even translators notes which can all help you create the best translation possible.
  
* POEdit will automatically generate a `.mo` file for every `.po` file when you save the file, so you don't need to worry about that.
  
* Next you need to make sure that the `.po` (and `.mo`) files you created have the correct file name and are in the correct place. WordPress is quite particular about this, so you need to get this right:
    - If you are translating a plugin - and in this case TGMPA behaves like a plugin -, the file needs to be called `{plugin-textdomain}-{locale}.po/.mo`. The textdomain is normally what the `.pot` file is called - in this case `tgmpa` and the locale is a language_country code combination. For Dutch that could be for instance `nl_NL` or `nl_BE`. So your file names will then become `tgmpa-nl_NL.po` and `tgmpa-nl_NL.mo`.
     Plugin translation files can be shipped with a plugin in, for instance, the `/wp-content/plugins/my-plugin/languages/` subdirectory or can be downloaded directly from WP.org for plugins which are in the WP.org repository in which case they will be in the `/wp-content/languages/plugins/` directory.
     
   - If you are translating a theme, the file names depend on where the file will be placed. If it is placed in the `/wp-content/themes/my-theme/languages/` directory, the files should be called `{locale}.po/.mo`, i.e. `nl_NL.po/.mo`. If the file will be placed in the `/wp-content/languages/themes/` directory, the file has to be called `{theme-textdomain}-{locale}.po/.mo`.
     
   - The actual subdirectory name within a plugin or theme can vary, some call it `locale`, others `lang` or `languages`. The name itself in this case does not matter, as long as all the translations are in the same directory.
     
* Lastly, to bring it all together, the plugin or theme has to actually *load* the textdomain using either one of the following calls:
   ```php
   load_plugin_textdomain( 'plugin_textdomain', false, '/relative/path/to/lang/folder/from/plugins-dir/' )
   load_theme_textdomain( 'my_theme' );
   ```
   As we've only just started down the path to making our own translation files available, that call is currently not included in TGMPA *yet*. We also would need to overcome the hurdle of themes in the wp.org repository not being allowed to have more than one text-domain, so for now, I'm hesitant to add this call to TGMPA.
   
  As you said you are working on a theme, to make it work for you for now, I would suggest adding the following code to the top of the TGMPA file, in your `functions.php` file or in the file containing your TGMPA configuration (`example.php`), whichever is easier:
   ```php
	/**
	 * Load translations for TGMPA.
	 */
	function load_textdomain() {
		load_theme_textdomain( 'tgmpa', get_template_directory() . '/tgm/languages' );
	}
	add_action( 'init', array( $this, 'load_textdomain' ), 1 );
   ```
   Once you've added that and put your translation in the right directory, you should see the translated text appearing.

Hope this helps and I look forward to your pull request !

Smile,
Juliette


For further reading:
https://codex.wordpress.org/I18n_for_WordPress_Developers
https://developer.wordpress.org/plugins/internationalization/localization/
https://developer.wordpress.org/plugins/internationalization/how-to-internationalize-your-plugin/
https://developer.wordpress.org/themes/functionality/localization/
https://developer.wordpress.org/themes/functionality/internationalization/

**Comment by geoclaps** [comment]
Hi @jrfnl ! First of all, thank you for the excellent comment on my pull! Much appreciated!

I am not that familiar with multilang themes and plugins so what you told me was pretty new to me.

I started again from the beginning and took your updated version of the .pot file, imported it into poedit and started translating. I now have a fully translated document in Dutch: tgmpa-nl_NL.po and tgmpa-nl_NL.mo. I than uploaded the files to the server (see screenshot) and loaded the translations dir into my theme via the suggested method. After this I am getting an error (see other screenshot).

I put the suggested code wich loads the translation in plugins.php (in my case the file that manages de plugins within my theme). This file is being included in the functions.php of my theme.
 
![php_error](https://cloud.githubusercontent.com/assets/7847695/12111724/d255320a-b397-11e5-8613-0fda0c7ffc4f.png)
![server](https://cloud.githubusercontent.com/assets/7847695/12111726/d62fdbd2-b397-11e5-9b75-fa8d41907174.png)

**Comment by jrfnl** [comment]
@geoclaps Sorry about that - my bad. I'm too used to always work with classes that I forgot to change the function name. Glad you managed to fix that, the function looks good.
I've update my original example for any future users finding this thread.



> Language is still not being set to Dutch :(

Now as for this, the solution is quite simple. Have a good look through my original reply and your own:

> I now have a fully translated document in Dutch: `tgmpa-nl_NL.po` and `tgmpa-nl_NL.mo`. I than uploaded the files to the server (see screenshot) 

From the screenshot I can see you put the files in a theme.

>> If it is placed in the `/wp-content/themes/my-theme/languages/` directory, the files should be called `{locale}.po/.mo`, i.e. `nl_NL.po/.mo`.

So, rename the files to `nl_NL.po/.mo` (without the `tgmpa-` prefix) and things should start working.


Additionally: silly question maybe, but you did set the site language to Dutch ? To check this: go to Settings -> General -> Language dropdown (bottom item on the page), make sure it's set to Dutch and save.

**Comment by geoclaps** [comment]
@jrfnl, no need to say sorry since you've helped out in this problem! 

I'm not really a backend programmer, just a front end developer and webdesigner so PHP is not my best area of expertise :) Glad you're helping me like this.

The problem is resolved. I misread the part in wich you stated that the lang files should only exist without the plugin prefix. In my case this was the last problem. All translations are now being imported via the mo file. Thank you so much!

If needed I can place the translation files onto github so other Dutch people can use the translated strings? If so, where do you want me to place them?

**Comment by jrfnl** [comment]
@geoclaps Glad to hear it's working now.

It would be great if you would send in a pull request with the translation files! Do you know how to do this or would you like me to talk you through it ?

**Comment by geoclaps** [comment]
If you could... that would be great. That way I know they are at the right place.

**Comment by jrfnl** [comment]
@geoclaps Ok, here goes.

* First off, you'll need to _fork_ this repository to your own account. You can do so by clicking the button at the top-right of this page.

* Next you need to add the two files.
   - I don't know whether you already have a Git client installed, if you have, clone your fork to your own computer, make sure you are on the `develop` branch, copy the two translation files to the `/languages/` sub-directory, commit the change and push the change to your fork on GitHub.
   
   - If you don't have a Git client installed (yet), you can either try and make the changes through the web interface ('New file' button), though I'm not sure how well that will work for the .mo file as it's a binary.
   Alternatively you can install a git client, like the [GH desktop application](https://desktop.github.com/) or - if you are on Windows - the excellent [GitExtensions](https://gitextensions.github.io/) application and then follow the above steps.

* Once you've pushed your changes up to GH, visit the GH page for your fork ( https://github.com/geoclaps/TGM-Plugin-Activation ), click the 'New pull request' button at the top of the file list and a form will open, describe the change you made and click 'Create pull request'.

**Pro-tip:**
If you get to a point where you regularly contribute to a project, I would advise against making changes in the `develop` or `master` branch, but suggest you create a new branch for each change and send a PR from that feature branch to the `develop` or `master` branch on the main repository. This allows for individual changes to be tested in isolation and will make managing your code changes easier.


For further reading:
https://desktop.github.com/
https://help.github.com/articles/set-up-git/
https://help.github.com/articles/fork-a-repo/

**Comment by jrfnl** [comment]
@geoclaps Hiya, just wondering how you're getting on with creating the PR. Need any more help ?

**Comment by jrfnl** [comment]
FYI: the translation loading calls for TGMPA have been merged into the `develop` branch in preparation of the  2.6.0 release.

Would be great if we could include the Dutch translation with that release! *nudge*

**Comment by geoclaps** [comment]
I'm sorry I haven't placed anymore posts in this topic since your last reply. To be honest, I forgot the entire matter.

I have the translated files at work so I can upload them to Github. First time for me doing this so I hope it will work out. I will keep you posted.

**Comment by jrfnl** [comment]
Thanks! i look forward to the pull request - let me know if you need help (and in the worst case, just mail me the files and I'll sort it out).

**Comment by jrfnl** [comment]
@geoclaps Any news ?

**Comment by geoclaps** [comment]
Yes, I had to translate the files again. I probably deleted the initial translations... :(

I have uploaded the po and the mo file. Hope they work!

**Comment by jrfnl** [comment]
Great! - can you turn that into a pull request please ?

If you go to the main page of your fork of TGMPA on GH, you will probably see a 'you are 1 commit ahead' kind of line with to the right a green button 'create PR'. If you click on that, that will open up a screen to open a new PR here which will allow us to merge your files into our copy.

**Comment by marcellovocale** [comment]
Hi @jrfnl 
with so i get a error
```
/**
 * Load translations for TGMPA.
 */
function theme_prefix_load_textdomain() {
    load_theme_textdomain( 'tgmpa', get_template_directory() . '/feature/include/required_plugin/TGM-Plugin-Activation/languages' );
}
add_action( 'init', array( $this, 'theme_prefix_load_textdomain' ), 1 );
```

> Warning: call_user_func_array() expects parameter 1 to be a valid callback, first array member is not a valid class name or object in C:\xampp\htdocs\wordpress3\wp-includes\class-wp-hook.php on line 298


but with so working:

```
/**
 * Load translations for TGMPA.
 */
function theme_prefix_load_textdomain() {
    load_theme_textdomain( 'tgmpa', get_template_directory() . '/feature/include/required_plugin/TGM-Plugin-Activation/languages' );
}
add_action( 'init', 'theme_prefix_load_textdomain' );
```

Can u please confirm me if this is the right way for load the language?Thx

**Comment by GaryJones** [comment]
This is a generic coding issue, not related to loading a textdomain.

The first approach (with the array and $this), is when then callback is actually a method on a class. The `add_action` call should be inside another method (constructor or an init method).

The second snippet is for a procedural (standalone) function.

**Comment by dejudicibus** [comment]
Well, I developed a theme called `sandbox`, which contains a languages directory where I have `it_IT.po` and `it_IT.mo` files for my source code. I created a `services` folder where I placed `class-tgm-plugin-activation.php`. It works fine. Now I want to use Italian for TGM too. In the TGM package there are  two files: `tgmpa-it_IT.po` and `tgmpa-it_IT.mo`. Where should I place them?

In my functions.php I have 

`load_theme_textdomain('sandbox', get_template_directory() . '/languages');`

Should I add anything for TGM too?
Thank you in adavance.
